src/core/files.ml

```diff
@@ -142,56 +142,64 @@ module ModMap :
-    terminated with a graceful error message. *)
-let set_lib_root : string -> unit = fun path ->
-  try
-    let path = Filename.realpath path in
-    if not (Sys.is_directory path) then
-      fatal_no_pos "Invalid library root: [%s] is not a directory." path;
-    match Stdlib.(!lib_root) with
-    | None    -> Stdlib.(lib_root := Some(path))
-    | Some(_) -> fatal_no_pos "The library root was already set."
-  with Sys_error(_) | Invalid_argument(_) ->
-    fatal_no_pos "Invalid library root: no such file or directory [%s]." path
-
-(** [default_lib_root ()] returns the default library root curently configured
-    for the system. It depends on the current Opam switch (if any), and it may
-    or may not correspond to an existing directory. *)
-let default_lib_root : unit -> string = fun _ ->
-  let prefix =
-    try Sys.getenv "OPAM_SWITCH_PREFIX"
-    with Not_found -> "/usr/local"
-  in
-  Filename.concat prefix "lib/lambdapi/lib_root"
-
-(** [lib_root_path ()] returns the canonical library root path. It corresponds
-    to a valid path to a directory. The function fails gracefully if it is not
-    able to find such a directory. Note that prior to installation, the option
-    ["--lib-root"] must be used so that this function does not fail. *)
-let lib_root_path : unit -> string = fun _ ->
-  match Stdlib.(!lib_root) with Some(path) -> path | None ->
-  let path = default_lib_root () in
-  try
-    let path = Filename.realpath path in
-    if not (Sys.is_directory path) then
-      fatal_no_pos "Invalid default library root [%s]: this is not a \
-                    directory." path;
-    path
-  with Sys_error(_) | Invalid_argument(_) ->
-    fatal_no_pos "Default library root [%s] does not exist." path
-
 (** [lib_mappings] stores the specified mappings of library paths. *)
 let lib_mappings : ModMap.t ref =
   ref ModMap.empty
 
-(** [init_lib_root ()] registers the currently set library root as part of our
-    module mapping. This function MUST be called before one can consider using
-    [module_to_file] or [module_path]. *)
-let init_lib_root : unit -> unit = fun _ ->
-  let root = lib_root_path () in
-  lib_mappings := ModMap.set_root root !lib_mappings
+(** [set_lib_root path] sets the library root. The following paths are
+    set sequentially, such that the active one is the last valid path
+    - [/usr/local/lib/lambdapi/lib_root/]
+    - [$OPAM_SWITCH_PREFIX/lib/lambdapi/lib_root]
+    - [$LAMBDAPI_LIB_ROOT/lib/lambdapi/lib_root]
+    - the path given as parameter of the [--lib-root] command line argument;
+      if the path given on command line is not a valid (existing) directory,
+      the program terminates with a graceful error message. *)
+let set_lib_root : string option -> unit = fun path ->
+  let open Stdlib in
+  let prefix = Stdlib.ref "/usr/local" in
+  begin match Sys.getenv_opt "OPAM_SWITCH_PREFIX" with
+    | Some(pth) -> prefix := pth
+    | None -> ()
+  end;
```

Comment:
here and later, use Option.iter instead ?

Reply:
Very good idea.